From 0e00aefe77e51d5ac3473c4252f1271de7bc510c Mon Sep 17 00:00:00 2001
From: Benjamin Berg <bberg@redhat.com>
Date: Sun, 6 Feb 2022 19:26:38 +0100
Subject: [PATCH 7/7] adaptive: Read the lid state directly from the kernel
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

The UPower interface for the lid state is deprecated. As such, we should
be reading it directly from the input device. Making the switch is
straight forward, so just do it.

Closes: #340
---
 src/thd_engine_adaptive.cpp | 43 ++++++++++++++++++++++++++-----------
 src/thd_engine_adaptive.h   |  1 +
 2 files changed, 32 insertions(+), 12 deletions(-)

diff --git a/src/thd_engine_adaptive.cpp b/src/thd_engine_adaptive.cpp
index da6a2e6..3e8b367 100644
--- a/src/thd_engine_adaptive.cpp
+++ b/src/thd_engine_adaptive.cpp
@@ -94,8 +94,18 @@ void cthd_engine_adaptive::destroy_dynamic_sources() {
 	if (tablet_dev) {
 		close(libevdev_get_fd(tablet_dev));
 		libevdev_free(tablet_dev);
+
+		if (lid_dev == tablet_dev)
+			lid_dev = NULL;
 		tablet_dev = NULL;
 	}
+
+	if (lid_dev) {
+		close(libevdev_get_fd(lid_dev));
+		libevdev_free(lid_dev);
+
+		lid_dev = NULL;
+	}
 }
 
 cthd_engine_adaptive::~cthd_engine_adaptive() {
@@ -848,7 +858,7 @@ int cthd_engine_adaptive::verify_condition(struct condition condition) {
 			|| condition.condition == (adaptive_condition) 0) {
 		return 0;
 	}
-	if (condition.condition == Lid_state && upower_client != NULL)
+	if (condition.condition == Lid_state && lid_dev != NULL)
 		return 0;
 	if (condition.condition == Power_source && upower_client != NULL)
 		return 0;
@@ -1010,11 +1020,16 @@ int cthd_engine_adaptive::evaluate_temperature_condition(
 
 int cthd_engine_adaptive::evaluate_lid_condition(struct condition condition) {
 	int value = 0;
-	bool lid_closed = up_client_get_lid_is_closed(upower_client);
 
-	if (!lid_closed)
-		value = 1;
+	if (lid_dev) {
+		struct input_event ev;
 
+		while (libevdev_has_event_pending(lid_dev))
+			libevdev_next_event(lid_dev, LIBEVDEV_READ_FLAG_NORMAL, &ev);
+
+		int lid_closed = libevdev_get_event_value(lid_dev, EV_SW, SW_LID);
+		value = !lid_closed;
+	}
 	return compare_condition(condition, value);
 }
 
@@ -1547,10 +1562,9 @@ void cthd_engine_adaptive::setup_input_devices() {
 	struct dirent **namelist;
 	int i, ndev, ret;
 
-	tablet_dev = NULL;
-
 	ndev = scandir("/dev/input", &namelist, is_event_device, versionsort);
 	for (i = 0; i < ndev; i++) {
+		struct libevdev *dev = NULL;
 		char fname[267];
 		int fd = -1;
 
@@ -1558,16 +1572,21 @@ void cthd_engine_adaptive::setup_input_devices() {
 		fd = open(fname, O_RDONLY | O_NONBLOCK | O_CLOEXEC);
 		if (fd < 0)
 			continue;
-		ret = libevdev_new_from_fd(fd, &tablet_dev);
+		ret = libevdev_new_from_fd(fd, &dev);
 		if (ret) {
 			close(fd);
 			continue;
 		}
-		if (libevdev_has_event_code(tablet_dev, EV_SW, SW_TABLET_MODE))
-			return;
-		libevdev_free(tablet_dev);
-		tablet_dev = NULL;
-		close(fd);
+
+		if (!tablet_dev && libevdev_has_event_code(dev, EV_SW, SW_TABLET_MODE))
+			tablet_dev = dev;
+		if (!lid_dev &&  libevdev_has_event_code(dev, EV_SW, SW_LID))
+			lid_dev = dev;
+
+		if (lid_dev != dev && tablet_dev != dev) {
+			libevdev_free(dev);
+			close(fd);
+		}
 	}
 }
 
diff --git a/src/thd_engine_adaptive.h b/src/thd_engine_adaptive.h
index d044698..fb55aa4 100644
--- a/src/thd_engine_adaptive.h
+++ b/src/thd_engine_adaptive.h
@@ -155,6 +155,7 @@ protected:
 	UpClient *upower_client;
 	GDBusProxy *power_profiles_daemon;
 	struct libevdev *tablet_dev;
+	struct libevdev *lid_dev;
 	int current_condition_set;
 	int policy_active;
 	int fallback_id;
-- 
2.35.1

